velox/exec/tests/utils/QueryAssertions.cpp

```diff
@@ -703,9 +775,135 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
+    std::vector<MaterializedRow>& data) {
+  for (auto i = 0; i < data.size(); ++i) {
+    velox::column_index_t lastNonFloatColumn = width - 1;
+    for (auto j = (int32_t)floatColumnIndices.size() - 1; j >= 0; --j) {
+      if (floatColumnIndices[j] != lastNonFloatColumn) {
+        std::swap(data[i][floatColumnIndices[j]], data[i][lastNonFloatColumn]);
+      }
+      --lastNonFloatColumn;
+    }
+  }
+}
+
+// Return true if currentRow and previousRow are the same at the first width
+// columns.
+bool equalPrefix(
+    const MaterializedRow& currentRow,
+    const MaterializedRow& previousRow,
+    size_t width) {
+  for (auto i = 0; i < width; ++i) {
+    if (currentRow[i] != previousRow[i]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Return true if there are multiple rows with the same prefix of the first
+// width columns. The argument groupedRows should be sorted by the first
+// width columns.
+bool hasMultiRowGroups(
+    const std::vector<MaterializedRow>& sortedRows,
+    size_t width) {
+  for (auto i = 1; i < sortedRows.size(); ++i) {
+    if (equalPrefix(sortedRows[i], sortedRows[i - 1], width)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+std::vector<velox::column_index_t> getFloatColumnIndices(const MaterializedRowMultiset& left,
+    const MaterializedRowMultiset& right) {
+  std::vector<velox::column_index_t> floatColumnIndices;
+  auto& leftRow = *left.begin();
+  auto& rightRow = *right.begin();
+  auto width = leftRow.size();
+  for (auto i = 0; i < width; ++i) {
+    if (leftRow[i].kind() != rightRow[i].kind()) {
+      VELOX_FAIL("Expect the same type between two result sets.");
+    }
+    if ((leftRow[i].kind() == TypeKind::REAL ||
+         leftRow[i].kind() == TypeKind::DOUBLE)) {
+      floatColumnIndices.push_back(i);
+    }
+  }
+  return floatColumnIndices;
+}
+
+// Sort rows in left and right into leftSorted and rightSorted
+// and returns whether the sorting by unique keys is successful. The sorting is
+// successful if the following requirements are satisfied.
+//   1. left and right are non-empty and have the same number of columns of the
+//   same types.
+//   2. Rows in left and right are grouped by non-floating-point column. If
+//   there are one or more floating-point columns, either left or right must
+//   have a single row in every group. This is not required if there is no
+//   floating-point column.
+// The sorted results are stored in leftSorted and rightSorted with the
+// floating-point columns moved to the end and rows in ascending order.
+bool sortByUniqueKey(
+    const MaterializedRowMultiset& left,
+    const MaterializedRowMultiset& right,
+    std::vector<MaterializedRow>& leftSorted,
+    std::vector<MaterializedRow>& rightSorted) {
+  if (left.empty() || right.empty() ||
+      left.begin()->size() != right.begin()->size()) {
+    return false;
+  }
+
+  auto floatColumnIndices = getFloatColumnIndices(left, right);
+  if (floatColumnIndices.empty()) {
```

Comment:
This logic doesn't seem to be described in the function comment. I'm thinking that this function is doing too many things. It might be cleaner to let it just do sorting and take list of floating point columns as input.

I'm also wondering whether the logic of moving floating point columns to the end would result in confusing error messages where missing and extra rows will be printed in an "unexpected" order that doesn't match the order of columns in the query plan. Would it be cleaner to do the sorting without moving the columns?